libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,955 @@
+		case Type::Warning:
+			return Severity::Warning;
+	}
+	// Should never be reached.
+	return Severity::Error;
+}
+
+bool LanguageServer::compile(std::string const& _path)
+{
+	// TODO: optimize! do not recompile if nothing has changed (file(s) not flagged dirty).
+
+	auto const i = m_fileReader->sourceCodes().find(_path);
+	if (i == m_fileReader->sourceCodes().end())
+		return false;
+
+	m_compilerStack.reset();
+	m_compilerStack = make_unique<CompilerStack>(bind(&FileReader::readFile, ref(*m_fileReader), _1, _2));
+
+	// TODO: configure all compiler flags like in CommandLineInterface (TODO: refactor to share logic!)
+	OptimiserSettings settings = OptimiserSettings::standard(); // TODO: get from config
+	m_compilerStack->setOptimiserSettings(settings);
+	m_compilerStack->setParserErrorRecovery(false);
+	m_compilerStack->setRevertStringBehaviour(RevertStrings::Default); // TODO get from config
+	m_compilerStack->setSources(m_fileReader->sourceCodes());
+	m_compilerStack->setRemappings(m_remappings);
+
+	m_compilerStack->setEVMVersion(m_evmVersion);
+
+	trace("compile: using EVM "s + m_evmVersion.name());
+
+	m_compilerStack->compile(CompilerStack::State::AnalysisPerformed);
+	return true;
+}
+
+void LanguageServer::validate(std::string const& _path)
+{
+	compile(_path);
+
+	Json::Value params;
+	params["uri"] = toFileURI(_path);
+
+	params["diagnostics"] = Json::arrayValue;
+	for (shared_ptr<Error const> const& error: m_compilerStack->errors())
+	{
+		// Don't show this warning. "This is a pre-release compiler version."
+		if (error->errorId().error == 3805)
+			continue;
+
+		SourceReferenceExtractor::Message const message = SourceReferenceExtractor::extract(*error);
+
+		Json::Value jsonDiag;
+		jsonDiag["source"] = "solc";
+		jsonDiag["severity"] = int(toDiagnosticSeverity(error->type()));
+		jsonDiag["message"] = message.primary.message;
+
+		if (message.errorId.has_value())
+			jsonDiag["code"] = Json::UInt64{message.errorId.value().error};
+
+		jsonDiag["range"] = toJsonRange(
+			message.primary.position.line, message.primary.startColumn,
+			message.primary.position.line, message.primary.endColumn
+		);
+
+		for (SourceReference const& secondary: message.secondary)
+		{
+			Json::Value jsonRelated;
+			jsonRelated["message"] = secondary.message;
+			jsonRelated["location"]["uri"] = toFileURI((m_basePath / boost::filesystem::path(secondary.sourceName)).generic_string());
+			jsonRelated["location"]["range"] = toJsonRange(
+				secondary.position.line, secondary.startColumn,
+				secondary.position.line, secondary.endColumn
+			);
+			jsonDiag["relatedInformation"].append(jsonRelated);
+		}
+
+		params["diagnostics"].append(jsonDiag);
+	}
+
+	m_client->notify("textDocument/publishDiagnostics", params);
+}
+
+frontend::ASTNode const* LanguageServer::findASTNode(LineColumn _position, std::string const& _fileName)
+{
+	if (!m_compilerStack)
+		return nullptr;
+
+	frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(_fileName);
```

Comment:
What if it did not parse?

Reply:
`ast(...)` raises then, and any exception is caught in the lowest layer where the message is being handled and then an error reply is generated. So far for that. However, I think maybe each individual caller could catch that too in order produce a more meaningful response where it makes sense.

Reply:
It creates a `CompilerError`. The meaning of those is "someone abused the C++ interface to the compiler.". Please check for errors and the state of the stack instead.